### PR TITLE
fix: usr: PRES-611: Evict entries from cache when backing file changes

### PR DESCRIPTION
**Pull request by sakshia125**
Used last modified date to check if the entry in cache was the latest version of the backing file. Evicted entries holding outdated data.
Earlier, the cache wasn't being updated at all.

### CODE REVIEW

File: rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeper.java [hunk 1]
```diff
@@ -68,22 +72,23 @@ public BookKeeper(Configuration conf)
...
-        }
-        catch (ExecutionException e) {
-            log.error(String.format("Error updating block status for %s [%d %d] : %s", remotePath, startBlock, endBlock, Throwables.getStackTraceAsString(e)));
-            throw new TException(e);
-        }
+        md = fileMetadataCache.getIfPresent(remotePath);
 
+        //md will be null when 2 users try to update the file in parallel and both their entries are invalidated.
+        // TODO: find a way to optimize this so that the file doesn't have to be read again in next request (new data is stored instead of invalidation)
+        if( md==null || md.getLastModified() != lastModified ) {
```
Comment (stagraqubole) [inline rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeper.java]: nit: leave spaces before and after `==`

File: rubix-core/src/test/java/com/qubole/rubix/core/TestUpdatingofCache.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        log.info("All set to test");
+
+        // This should be after server comes up else client could not be created
+        inputStream = new CachingInputStream(fsDataInputStream, conf, backendPath, file.length(),file.lastModified(), new CachingFileSystemStats());
+    }
+
+    public void setup()
+            throws InterruptedException, IOException
+    {
+        final Configuration conf = new Configuration();
```
Comment (stagraqubole) [inline rubix-core/src/test/java/com/qubole/rubix/core/TestUpdatingofCache.java]: This `setup` methods seems unnecessary

File: rubix-core/src/test/java/com/qubole/rubix/core/TestUpdatingofCache.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        inputStream.seek(100);
+        byte[] buffer = new byte[1000];
+        int readSize = inputStream.read(buffer, 0, 1000);
+        String expectedOutput = DataGen.generateContent().substring(100, 1100);
+        assertions(readSize, 1000, buffer,expectedOutput);
+    }
+
+    @Test
+    public void testEviction()
+            throws IOException, InterruptedException
```
Comment (stagraqubole) [inline rubix-core/src/test/java/com/qubole/rubix/core/TestUpdatingofCache.java]: There is lot of code duplication here with only this method (testEviction) being new code. Move this test to TestCachingInputStream, that will save lot of duplication

File: rubix-core/src/test/java/com/qubole/rubix/core/TestUpdatingofCache.java (hunk 2 above)
Comment (sakshia125) [inline rubix-core/src/test/java/com/qubole/rubix/core/TestUpdatingofCache.java]: The initial setup starts the bookkeeper service and populates the backend file. We do not want these steps to repeat when we start a new CachingInputStream.

**Pull request closed without merging**